index.rst

```diff
@@ -48,14 +48,126 @@
+The execution system is already capable of over-writing an existing file if it is not known to registry and datastore ensures that the file name is unique for a specific dataId and run combination.
+
+If pre-emption occurs during the final registry import it is assumed that there is a single transaction that will be completely rolled back.
+The end result would be that the entire processing would have to be redone.
+If the pre-emption occurs in the short time between completion of the registry import and the completion of the job, we should ensure that the ``pipetask`` command is configured to complete without action when the job is restarted.
+
+Externalized
+------------
+
+The most flexible approach is for the workflow generator to insert extra jobs into the workflow that handle registry merging and registry syncing.
+
+A new job would be inserted before the initial pipeline jobs in the graph that downloads the initial registry state from the cloud registry and stores it in a SQLite file.
+This SQLite file would then be provided as an input to the first job.
+The pipeline job would use that SQLite registry and a GCS-with-local-cache datastore as described in the previous section.
+On completion the output collected by the workflow execution system would be an updated SQLite registry.
+If the next pipeline job only takes files from a single input job this SQLite file would be passed directly to the next pipeline job.
+If, on the other hand, the next job gathers files from multiple jobs, the workflow generator will insert a new job before it that reads all the SQLite files and merges them into a single output for that job.
+It can optionally also sync registry information to the cloud registry and trim the SQLite file that it passes on.
+This sync and prune could be a distinct job that can be inserted after the merge job in the graph.
+At the end node (or nodes) of each workflow graph a final job will be inserted that updates the cloud registry with the final state.
+
+In this scenario pre-emption has no impact on registry state for pipeline jobs since they are running with an entirely local SQLite registry.
+Pre-emption still has to be understood in terms of the registry sync jobs and requires that the remote update of the cloud registry happens in a single transaction.
+
+This approach allows the submission system to decide whether the registry is updated multiple times within the graph or solely at the end since the registry merging jobs can be configured to either merge the inputs and pass them on complete, or merge, sync and trim.
+A trimmed registry can not be passed on to the next job if the remainder has not been synced with the cloud registry.
+
+Summary
+=======
+
+Both the job-level pooling and fully externalized approach to Registry should be able to improve throughput significantly.
+The job-level pooling approach is simplest to implement and should be attempted first and performance compared to the current system.
```

Comment:
I think I agree that we should try job-level pooling first if we're worried specifically about being able to stand up the externalized approach in time for DC0.2.  But I think the lack of flexibility may make it a dead end, so if we think we can get enough of the externalized approach working by then, it's a better long-term bet.

Reply:
The job-level pooling is essentially independent of anything BPS or the graph builder is doing so in theory that could be implemented without having to worry about where the Workflow Graph boundary lives between pipe_base/ctrl_mpexec/ctrl_bps. I've been assuming that makes it simpler to start on. The other option would involve changes to more of the system and we should discuss if we want to do that as a bit of a "tiger team" group effort.

Reply:
I guess I don't understand job-level pooling.  If each job is still connecting to the central database at the beginning and end of the job, how does that help throttle the number of simultaneous connections to the database without throttling the jobs?  And what does job-level pooling get us until we have multiple Quanta per job?

Reply:
Pooling allows all the transactions for butler.put to happen locally so that they aren't kept open for a huge amount of extra time whilst the data are being uploaded. Doing the registry export at the start and import at the end means that you are keeping the registry connections alive for a significantly shorter period.